admob/src/include/firebase/admob/types.h

```diff
@@ -82,13 +92,142 @@ enum AdMobError {
+  kAdMobErrorServerError,
+  /// The current device’s OS is below the minimum required version.
+  kAdMobErrorOSVersionTooLow,
+  /// The request was unable to be loaded before being timed out.
+  kAdMobErrorTimeout,
+  /// Will not send request because the interstitial object has already been
+  /// used.
+  kAdMobErrorInterstitialAlreadyUsed,
+  /// The mediation response was invalid.
+  kAdMobErrorMediationDataError,
+  /// Error finding or creating a mediation ad network adapter.
+  kAdMobErrorMediationAdapterError,
+  /// Attempting to pass an invalid ad size to an adapter.
+  kAdMobErrorMediationInvalidAdSize,
+  /// Invalid argument error.
+  kAdMobErrorInvalidArgument,
+  /// Received invalid response.
+  kAdMobErrorReceivedInvalidResponse,
+  /// Will not send a request because the rewarded ad object has already been
+  /// used.
+  kAdMobErrorRewardedAdAlreadyUsed,
+  /// A mediation ad network adapter received an ad request, but did not fill.
+  /// The adapter’s error is included as an underlyingError.
+  kAdMobErrorMediationNoFill,
+  /// Will not send request because the ad object has already been used.
+  kAdMobErrorAdAlreadyUsed,
+  /// Will not send request because the application identifier is missing.
+  kAdMobErrorApplicationIdentifierMissing,
+  /// Android Ad String is invalid.
+  kAdMobErrorInvalidAdString,
   /// Fallback error for any unidentified cases.
   kAdMobErrorUnknown,
 };
+
 #ifdef INTERNAL_EXPERIMENTAL
 // LINT.ThenChange(//depot_firebase_cpp/admob/client/cpp/src_java/com/google/firebase/admob/internal/cpp/ConstantsHelper.java)
 #endif  // INTERNAL_EXPERIMENTAL
 
+/// Information about why an ad operation failed.
+class AdResult {
+ public:
+  /// Copy Constructor.
+  AdResult(const AdResult& ad_result);
+
+  /// Destructor.
+  virtual ~AdResult();
+
+  /// Assignment operator.
+  AdResult& operator=(const AdResult& obj);
+
+  /// If the operation was successful then the other error reporting methods
+  /// of this object will return defaults.
+  bool is_successful() const;
+
+  /// Retrieves an AdResult which represents the cause of this error.
+  ///
+  /// @return a pointer to an AdResult which represents the cause of this
+  /// AdResult.  If there was no cause, or if this result was successful,
+  /// then nullptr is returned.
+  std::unique_ptr<AdResult> GetCause() const;
+
+  /// Gets the error's code.
+  AdMobError code() const;
+
+  /// Gets the domain of the error.
+  const std::string& domain() const;
+
+  /// Gets the message describing the error.
+  const std::string& message() const;
+
+  /// Returns a log friendly string version of this object.
+  const std::string& ToString() const;
+
+  /// A domain string which represents an undefined error domain.
+  ///
+  /// The Admob SDK returns this domain for domain() method invocations when
+  /// converting error information from legacy mediation adapter callbacks.
+  static const char* kUndefinedDomain;
```

Comment:
nit: I think this needs to be a static const char* const, or a static const char[], otherwise the pointer can be reassigned.

Reply:
Done.

Reply:
Nice, thanks. Done!